### PR TITLE
Network resource

### PR DESCRIPTION
**Pull request by ueokande**
Network resource is supported.  Define Network resources as following:

```yaml
# cluster.yml
kind: Network
name: net1
spec:
  addresses:
    - 172.16.0.1/24
---
kind: Node
name: boot-0
spec:
  volumes:
    - name: system
      size: 10G
      recreatePolicy: Never
  interfaces:
    - net1
```

Placemat creates bridges (virtual L2 bridges) corresponding to Network resources.  The `addresses` parameter contains IP addresses which host assigns to the bridge.

You can observe a bridge named `net1`, and TAP named `boot-0_net1` which VM connets.  The VM can connecto to host's bridge with IP addresses `172.16.0.0/24`.

```console
$ ip link
1126: net1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether ea:43:bc:a8:bf:51 brd ff:ff:ff:ff:ff:ff
1127: boot-0_net1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast master net1 state UP mode DEFAULT group default qlen 1000
    link/ether ea:43:bc:a8:bf:51 brd ff:ff:ff:ff:ff:ff
```

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -35,24 +40,60 @@ func createNodeVolumes(ctx context.Context, provider Provider, cluster *Cluster)
...
-	env.Stop()
-	return env.Wait()
+}
+func handleDestroyNetwork(env *cmd.Environment, provider Provider, networks []*Network) {
+	names := make([]string, len(networks))
+	for i, n := range networks {
+		names[i] = n.Name
+	}
+	env.Go(func(ctx context.Context) error {
+		<-ctx.Done()
```
Comment (ueokande) [inline server.go]: envのcontextがcancelされた時に終了する。

**Comment by ymmt2005** [comment]
@ueokande 
How about IPv6 addresses?

File: qemu.go [hunk 2]
```diff
@@ -37,10 +85,37 @@ func (q QemuProvider) CreateVolume(ctx context.Context, node string, vol *Volume
...
 	params := []string{"-enable-kvm"}
+
+	for _, br := range n.Spec.Interfaces {
+		tap := n.Name + "_" + br
+		err := createTap(ctx, tap, br)
+		if err != nil {
+			return err
+		}
+
+		params = append(params, "-netdev", "tap,id="+br+",ifname="+tap+",script=no,downscript=no")
```
Comment (ymmt2005) [inline qemu.go]: `,vhost=on` を付けないと vhost-net が有効にはならないですが、意図的でしょうか？
カーネルに vhost_net モジュールが必要なので、GCE インスタンスにないとかかな？

**Comment by ueokande** [comment]
@ymmt2005 IPv6 is also supported because the placemat holds addresses as strings, and pass them to `ip addr add` command.

```
kind: Network
name: net1
spec:
  addresses:
    - 172.16.0.1/24
    - 1111:2222:3333:4444::/64
```

File: qemu.go (hunk 2 above)
Comment (ueokande) [reply qemu.go]: 有無をチェックして確認します。なおGCEは対応してませんでした...

File: cmd/placemat/yaml_test.go [hunk 3]
```diff
@@ -93,6 +142,9 @@ name: node2
 	if err != nil {
 		t.Error(err)
 	}
+	if len(cluster.Networks) != 1 {
+		t.Error("len(cluster.Networks) != 2, ", len(cluster.Networks))
```
Comment (dulltz) [inline cmd/placemat/yaml_test.go]: 1 だと思います

File: qemu.go [hunk 4]
```diff
@@ -26,6 +69,33 @@ func (q QemuProvider) VolumeExists(ctx context.Context, node, vol string) (bool,
...
+	for _, addr := range net.Spec.Addresses {
+		err := cmd.CommandContext(ctx, "ip", "addr", "add", addr, "dev", net.Name).Run()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// DestroyNetwork destroies a bridge by the name
```
Comment (dulltz) [inline qemu.go]: s/destroies/destroys/

**Comment by ymmt2005** [comment]
As my review comments were addressed, I approve this PR.
Please address @dulltz comments.

**Comment by ueokande** [comment]
conflicted... :sob::sob::sob:

**Comment by ueokande** [comment]
resolved

**Comment by dulltz** [comment]
LGTM :+1:

**Pull request merged**